x/programs/rust/sdk-macros/src/lib.rs

```diff
@@ -293,17 +293,15 @@ pub fn state_keys(_attr: TokenStream, item: TokenStream) -> TokenStream {
                         .enumerate()
                         .map(|(i, field)| Ident::new(&format!("field_{i}"), field.span()));
                     let fields_2 = fields.clone();
-                    let fields_3 = fields.clone();
 
                     Ok(quote! {
                         Self::#variant_ident(#(#fields),*) => {
-                            let len = 1 + #(#fields_2.as_ref().len())+*;
+                            let as_bytes = borsh::to_vec(&(#(#fields_2),*))?;
+                            let len = 1 + as_bytes.len();
                             let len = u32::try_from(len).map_err(|_| std::io::ErrorKind::InvalidData)?;
                             writer.write_all(&len.to_le_bytes())?;
                             writer.write_all(&[#idx])?;
-                            #(
-                                writer.write_all(#fields_3.as_ref())?;
-                            )*
+                            writer.write_all(&as_bytes)?;
```

Comment:
You can simplify this whole thing down to a single line:
```rust
borsh::to_vec(&(#idx, #(#fields_2),*))?.serialize(writer);
```

Reply:
I don't think that works because we need the length in order to borsh deserialize it on the host side.
If we keep it, then we are going to borsh serialize the bytes twice which is a loss of performance and borsh serialize the `#idx` which is not really needed because it's a one-way conversion

Reply:
You have the length: https://docs.rs/borsh/latest/src/borsh/ser/mod.rs.html#254-265

Reply:
Thanks, yeah that makes sense!